Marshallable

jarsen opened the pull request:
Changes `ObjectConvertible` to `Marshallable`, as well as adds `Unmarshallable` and `Remarshallable`

refs #8
refs #14

jarsen committed:
add Marshallable & Unmarshallable

ObjectConvertible is now Marhsallable. This also adds an Unmarshallable
protocol which doesn't do much but defines a protocol people can adopt.

fixes #14

jarsen committed:
correct doc header

jarsen committed:
add Remarshallable

refs #8

bwhiteley commented:
The names are backwards. 
Marshallable == serializable and Unmarshallable == deserializable.

brianmullen commented:
Since the names are backwards, `Remarshallable` doesn't make sense for updating. Maybe we combine the update method with `Unmarshallable` and add a default implementation that doesn't do anything?

jarsen commented:
Oh noes. I wonder if that means A) I'm dumb, B) it's not intuitive enough, or C) people just need to learn which is which.

re: `Remarshallable`, I think I would prefer to have a separate protocol. But that could work.

jarsen committed:
everything is backwards

jarsen commented:
The problem I guess with having an update function on `Unmarshallable` with a default implementation is that people don't have to implement it, and it might do nothing. I could see this being confusing. I say we just come up with a better name. `UnmarshallUpdatable`? Something along those lines.

brianmullen commented:
I don't know if I am a big fan of the name `Unmarshallable`. I think its the `Un` prefix as the first thing that pops into my head is undoing something. I know we are technically undoing the "serialized" data but in still seems confusing. What about changing the name to `Demarshallable`?

I am ok with a separate protocol for updating (I would prefer it). Just with the current name of `Remarshallable` didnt make sense.

jarsen commented:
Right, I agree `Resmarshallable` no longer makes sense. I do find `Unmarshallable` more confusing as well—because it could mean it can not be marshalled. How about `Unmarshalling` and `Marshalling` instead?

jarsen commented:
And `UnmarshalUpdating`?

jarsen commented:
Also, for the sake of consistency with the framework name, I think I should change everything to me marshal with one "L". It looks like it is technically correct, but marshaling is just a weird word that can be spelled two ways (first sentence https://en.wikipedia.org/wiki/Marshalling_%28computer_science%29)

jarsen committed:
"ing"s and single "L"s

`Marshallable` -> `Marshaling`
`Unmarshallable` -> `Unmarshaling`
`Remarshallable` -> `MarshalUpdating`

jarsen committed:
whoops. finish what I started

jarsen commented:
Here's a funny thing, I renamed `Unmarshaling.value` to `Unmarshaling.unmarshall`, and in doing so I actually broke it without realizing. Still conforms to `ValueType` but gets the default implementation :astonished:

jarsen committed:
Fix bug.

Can't wait for tests!

bwhiteley commented on Marshal/Marshaling.swift:
```diff
@@ -0,0 +1,17 @@
+
+public protocol Marshaling {
+    func marshal() -> Object
+}
```
Small nit: no newline at end of file.

jarsen replied:
good catch.

bwhiteley commented on Marshal/UnmarshalUpdating.swift:
```diff
@@ -0,0 +1,19 @@
+import Foundation
+
+
+public protocol UnmarshalUpdating {
```
Small nit: vertical whitespace is inconsistent with Marshaling.swift (above and below the `import`)

jarsen replied:
I actually have noticed this and it has been driving me nuts. Saving it for a different commit or PR

bwhiteley commented:
:+1:

bwhiteley commented:
:+1:

jarsen committed:
fix minor styling issues

not all files ended with a new line, and spacing between the awesome
doc header and imports was incosistent

jarsen commented:
fixes #8 
fixes #14
